src/compiler/scala/tools/nsc/ast/TreeGen.scala

```diff
@@ -366,15 +367,21 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
 
   def expandFunction(localTyper: analyzer.Typer)(fun: Function, inConstructorFlag: Long): Tree = {
     val anonClass = fun.symbol.owner.newAnonymousFunctionClass(fun.pos, inConstructorFlag)
-    val parents = if (isFunctionType(fun.tpe)) {
-      anonClass addAnnotation SerialVersionUIDAnnotation
-      addSerializable(abstractFunctionType(fun.vparams.map(_.symbol.tpe), fun.body.tpe.deconst))
-    } else {
-      if (fun.tpe.typeSymbol.isSubClass(SerializableClass))
-        anonClass addAnnotation SerialVersionUIDAnnotation
-      fun.tpe :: Nil
+    val typeSym = fun.tpe.typeSymbol
+    val isFunction = isFunctionSymbol(typeSym)
+    if (isFunction || typeSym.isSubClass(SerializableClass))
+      anonClass.addAnnotation(SerialVersionUIDAnnotation)
+
+    val rScope = newScope
+    def parents(tp: Type): List[Type] = tp match {
+      case RefinedType(ps, scope) =>
+        assert(scope.forall(_.isType), s"Cannot expand function of type $tp")
+        ps.flatMap(parents).tap(_ => scope.foreach(rScope.enter))
```

Comment:
Any reason to use `tap` here instead of a statement above?

Reply:
To preserve the order, but I assume it doesn't matter in any way

Reply:
I find it elegant, and inlines well under `-opt:inline:scalaworld`.